### ISSUE TITLE
Limit of 2G ignored while uploading files.

### ISSUE THREAD

**Issue by nicolasfranck** [ugent-library/deliver#38]
e.g. if you set the maximum bytes on this [line](https://github.com/ugent-library/deliver/blob/main/controllers/folders.go#L136) to `2000`, then it will not be respected. Something to do with the rewrite of the controller interface?

If you inspect the request body directly you'll see also that the request body seems to be empty, while the form file does contain something. Which is odd, as the file is sent IN the request body, and that check is performed BEFORE checking the form file. Even if I fully exhaust the request body, the form file still works afterwards. So I guess that form file is using some different request body?

**Comment by nicolasfranck** [comment]
Possible trace in middleware [CSRF](https://github.com/gorilla/csrf): `Note that the CSRF middleware will (by necessity) consume the request body if the token is passed via POST form values. If you need to consume this in your handler, insert your own middleware earlier in the chain to capture the request body.`

In biblio-backoffice the csrf token is read from the page meta tag `csrf-token` and added to every request header when issuing an htmx request.
So the csrf middleware does not feel the need to read the request body?

**Comment by nics** [comment]
good find, we can the same in the ajax call

**Comment by nicolasfranck** [comment]
one possible solution, as @nics noted: create a middleware that implements the MaxBytesReader, and put as the first middleware in front of everything else. That way it doesn't matter.

**Comment by nicolasfranck** [comment]
Important note: due to the way forms are handled server side, sending a file within a form (with `multipart/form-data`) almost always results in a temporary file (almost, i.e. if size is bigger than 32M). That explains the high upload throughput, and the long silence when the progress bar reaches 100%.

I did a separate experiment, where I tried to post the file contents directly inside the request body, without form encoding, and by sending the file attributes in the request headers. That way I can feed the request body directly to S3. Results: upload goes a bit slower, but takes less time at the end. This suggests that the contents are streamed directly into S3, instead of streaming to disk at first.
